Sources/TensorUtils/Math.swift

```diff
@@ -76,27 +75,38 @@ public struct Math {
     /// and their softmaxed probabilities.
     /// 
     public static func topK(arr: [Float], k: Int) -> (indexes: [Int], probs: [Float]) {
-        var minV = -Float.greatestFiniteMagnitude
-        var selected: [(index: Int, value: Float)] = []
-
-        for (i, v) in arr.enumerated() {
-            if v > minV || selected.count < k {
-                // Append and sort
-                if selected.count == k {
-                    selected.remove(at: 0)
-                }
-                selected.append((i, v))
-                selected.sort { $0.value < $1.value }
-                minV = selected.first!.value
-            }
+        guard !arr.isEmpty else {
+            return (indexes: [], probs: [])
         }
-
-        selected = selected.reversed()
-        let indexes = selected.map { $0.index }
-        let logits = selected.map { $0.value }
-        let probs = softmax(logits)
-
-        return (indexes: indexes, probs: probs)
+        let k = min(k, arr.count)
+        let arrDescriptor = BNNSNDArrayDescriptor.allocate(
+            initializingFrom: arr,
+            shape: .vector(arr.count)
+        )
+        let bestIndices = BNNSNDArrayDescriptor.allocateUninitialized(
+            scalarType: Int32.self,
+            shape: .vector(k)
+        )
+        let bestValues = BNNSNDArrayDescriptor.allocateUninitialized(
+            scalarType: Float.self,
+            shape: .vector(k)
+        )
+        try? Accelerate.BNNS.applyTopK(
+            k: k,
+            input: arrDescriptor,
+            bestValues: bestValues,
+            bestIndices: bestIndices,
+            axis: 0,
+            batchSize: 1,
+            filterParameters: nil
+        )
+        let distances = bestValues.data!.withMemoryRebound(to: Float.self, capacity: k) { ptr in
```

Comment:
Is there anything sensible to do if the operation fails rather than crashing? I'm fine merging as is, just wandering if you have any ideas that don't complicate things too much.

Reply:
I'm ok with merging it as is for now as well. Documentation is rather silent about the error being thrown, so I suspect that the swift function definition might be autogenerated from some lower level C code. Making this function throwable might just overcomplicate it too much with not much of a gain.

Reply:
@pcuenca  do you think it'd be better to use `try!` here?

Reply:
Nah, I guess it will fail in the following line anyway.

Reply:
Actually, yes, let's do that just in case.

Reply:
done, forgot to call `deallocate` on `BNNSNDArrayDescriptor` so added it as well